pkg/storage/gc_queue.go

```diff
@@ -767,30 +774,50 @@ func RunGC(
 				startIdx := 1
 				if meta.Txn != nil {
 					// Keep track of intent to resolve if older than the intent
-					// expiration threshold.
-					if meta.Timestamp.Less(intentExp) {
+					// expiration threshold. Don't do this if we've already collected
+					// a lot of potential intents (note that what we collect here is
+					// extremely likely to be collectable unless the user is running
+					// very large numbers of very long transactions). But the same
+					// consideration as the TODO below holds: we should be handling
+					// the chunk and resetting the budget.
+					if intentBytes <= intentBudgetBytes && meta.Timestamp.Less(intentExp) {
 						txnID := meta.Txn.ID
 						txn := &roachpb.Transaction{
 							TxnMeta: *meta.Txn,
 						}
 						txnMap[txnID] = txn
 						infoMu.IntentsConsidered++
 						intentSpanMap[txnID] = append(intentSpanMap[txnID], roachpb.Span{Key: expBaseKey})
+						intentBytes += len(expBaseKey)
 					}
 					// With an active intent, GC ignores MVCC metadata & intent value.
 					startIdx = 2
 				}
+				if keyBytes > keyBudgetBytes {
+					// Stop considering keys for GC if we've already crashed the budget.
+					//
+					// TODO(tschottdorf): we already send multiple GCRequests if we have too many keys,
+					// but we should be dispatching them as we scan to avoid holding the whole chunk of
+					// keys in memory. Large Ranges should be the exception, but they can definitely be
+					// configured and even if not, they could conceivably exist.
+					return
+				}
 				// See if any values may be GC'd.
 				if gcTS := gc.Filter(keys[startIdx:], vals[startIdx:]); gcTS != (hlc.Timestamp{}) {
-					// TODO(spencer): need to split the requests up into
-					// multiple requests in the event that more than X keys
-					// are added to the request.
 					gcKeys = append(gcKeys, roachpb.GCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
+					keyBytes += len(expBaseKey)
 				}
 			}
 		}
 	}
 
+	if intentBytes > intentBudgetBytes {
```

Comment:
Will this... work? We haven't called `processKeysAndValues` yet.

Reply:
Yeah, I think this needs a test.